### DISCUSSION TITLE
Netty consumes all server memory in many DirectByteBuffer objects

### DISCUSSION THREAD

**Discussion by dpetukhov** [netty/netty#14129] [General]
Hi.

I'm debugging server running in production. It uses netty. And as it runs, it consumes memory untill all the server memory is over.  I noticed that memory consumed in DirectByteBuffer:

![image](https://github.com/netty/netty/assets/946425/6e58a6fb-dbd4-4376-b149-aaad04afb896)

and when server runs, the count of these buffers always increase.

Please suggest how to debug the issue?
I tried `-XX:MaxDirectMemorySize=1G`, but I get OutOfMemory exceptions

**Comment by franz1981** [comment]
Netty is using off-heap pools which, similarly to malloc arenas, grows to accomodate the required capacity at peak usage (i.e. not free yet allocated memory): if you have some high load the off-heap pools will grow to accomodate for it, reaching some steady-state or just throwing OOM (like malloc), but in a healthy system, it shouldn't happen i.e. a system where the load is correctly flow controlling the request/responses and the configured amount of memory is ok.

BUT...

if the memory keep on growing over and over without reaching a steady state, maybe you have a leak, and I suggest to read this: https://netty.io/wiki/reference-counted-objects.html#troubleshooting-buffer-leaks

**Comment by normanmaurer** [comment]
Also you should check where the reference to these buffers are held.
